`ArrayIndexOutOfBoundsException` error for multi-stage query

We're hitting an error here: https://github.com/apache/pinot/blob/master/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/OpChainSchedulerService.java#L80 on a multi-stage query. How can I debug this further? We're running on [this commit](https://github.com/apache/pinot/commit/e35f43c151bc0da540ad92324bb4771be509ae1b) on master for this cluster.

```
[2025-02-24 18:09:13.887616] INFO [DataBlockUtils] [query-runner-on-8442-1-thread-315792:52] Caught exception while processing query
[2025-02-24 18:09:13.887623] java.lang.RuntimeException: Cancelled by sender with exception: Error block {1000=null
[2025-02-24 18:09:13.887628] java.lang.ArrayIndexOutOfBoundsException}
[2025-02-24 18:09:13.887645] 	at org.apache.pinot.query.mailbox.GrpcSendingMailbox.cancel(GrpcSendingMailbox.java:108) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887659] 	at org.apache.pinot.query.runtime.operator.exchange.BlockExchange.cancel(BlockExchange.java:137) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887673] 	at org.apache.pinot.query.runtime.operator.MailboxSendOperator.cancel(MailboxSendOperator.java:189) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887685] 	at org.apache.pinot.query.runtime.operator.OpChain.cancel(OpChain.java:96) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887699] 	at org.apache.pinot.query.runtime.executor.OpChainSchedulerService$1.runJob(OpChainSchedulerService.java:82) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887711] 	at org.apache.pinot.core.util.trace.TraceRunnable.run(TraceRunnable.java:40) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887719] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
[2025-02-24 18:09:13.887727] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
[2025-02-24 18:09:13.887735] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
[2025-02-24 18:09:13.887742] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
[2025-02-24 18:09:13.887748] 	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
[2025-02-24 18:09:13.887754] Caused by: java.lang.RuntimeException: Error block {1000=null
[2025-02-24 18:09:13.887765] java.lang.ArrayIndexOutOfBoundsException}
[2025-02-24 18:09:13.887778] 	at org.apache.pinot.query.runtime.executor.OpChainSchedulerService$1.runJob(OpChainSchedulerService.java:80) ~[pinot-all-1.3.0-2025-01-13-997d04630b-SNAPSHOT-jar-with-dependencies.jar:1.3.0-2025-01-13-997d04630b-SNAPSHOT-997d04630b460e2e02b679fc63d8fe0f27610e0b]
[2025-02-24 18:09:13.887782] 	... 6 more
```

cc @Jackie-Jiang

--- a/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java
+++ b/pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java
@@ -1257,30 +1257,26 @@ public class MultiStageEngineIntegrationTest extends BaseClusterIntegrationTestS
       throws Exception {
     // Query written this way with a CTE and limit will be planned such that the multi-stage group by executor will be
     // used for both leaf and final aggregation
-    String sqlQuery = "SET mseMaxInitialResultHolderCapacity = 1;\n"
-        + "WITH tmp AS ("
-        + "  SELECT *"
-        + "  FROM mytable"
-        + "  WHERE AirlineID > 20000"
-        + "  LIMIT 10000"
-        + ") "
-        + "SELECT AirlineID,"
-        + "  COUNT(*) FILTER ("
-        + "    WHERE Origin = 'garbage'"
-        + ") "
-        + "FROM tmp "
-        + "GROUP BY AirlineID";
-    JsonNode result = postQuery(sqlQuery);
-    assertNoError(result);
-    // Ensure that result set is not empty
-    assertTrue(result.get("numRowsResultSet").asInt() > 0);
+    String aggregates1 = "COUNT(*) FILTER (WHERE Origin = 'garbage')";
+    String aggregates2 = aggregates1 + ", COUNT(*)";
+    String queryTemplate = "SET mseMaxInitialResultHolderCapacity = 1;\n"
+        + "WITH tmp AS (SELECT * FROM mytable WHERE AirlineID > 20000 LIMIT 10000)\n"
+        + "SELECT AirlineID, %s FROM tmp GROUP BY AirlineID";
+    String query1 = String.format(queryTemplate, aggregates1);
+    String query2 = String.format(queryTemplate, aggregates2);
+    for (String query : new String[]{query1, query2}) {
+      JsonNode result = postQuery(query);
+      assertNoError(result);
+      // Ensure that result set is not empty
+      assertTrue(result.get("numRowsResultSet").asInt() > 0);
 
-    // Ensure that the count is 0 for all groups (because the aggregation filter does not match any rows)
-    JsonNode rows = result.get("resultTable").get("rows");
-    for (int i = 0; i < rows.size(); i++) {
-      assertEquals(rows.get(i).get(1).asInt(), 0);
-      // Ensure that the main filter was applied
-      assertTrue(rows.get(i).get(0).asInt() > 20000);
+      // Ensure that the count is 0 for all groups (because the aggregation filter does not match any rows)
+      JsonNode rows = result.get("resultTable").get("rows");
+      for (int i = 0; i < rows.size(); i++) {
+        assertEquals(rows.get(i).get(1).asInt(), 0);
+        // Ensure that the main filter was applied
+        assertTrue(rows.get(i).get(0).asInt() > 20000);
+      }
     }
   }
 

--- a/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultistageGroupByExecutor.java
+++ b/pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultistageGroupByExecutor.java
@@ -71,16 +71,9 @@ public class MultistageGroupByExecutor {
   // because they use the zero based integer indexes to store results.
   private final GroupIdGenerator _groupIdGenerator;
 
-  public MultistageGroupByExecutor(
-      int[] groupKeyIds,
-      AggregationFunction[] aggFunctions,
-      int[] filterArgIds,
-      int maxFilterArgId,
-      AggType aggType,
-      boolean leafReturnFinalResult,
-      DataSchema resultSchema,
-      Map<String, String> opChainMetadata,
-      @Nullable PlanNode.NodeHint nodeHint) {
+  public MultistageGroupByExecutor(int[] groupKeyIds, AggregationFunction[] aggFunctions, int[] filterArgIds,
+      int maxFilterArgId, AggType aggType, boolean leafReturnFinalResult, DataSchema resultSchema,
+      Map<String, String> opChainMetadata, @Nullable PlanNode.NodeHint nodeHint) {
     _groupKeyIds = groupKeyIds;
     _aggFunctions = aggFunctions;
     _filterArgIds = filterArgIds;
@@ -248,10 +241,7 @@ public class MultistageGroupByExecutor {
     return rows;
   }
 
-  private Object[] getRow(
-      Iterator<GroupIdGenerator.GroupKey> groupKeyIterator,
-      int numKeys,
-      int numFunctions,
+  private Object[] getRow(Iterator<GroupIdGenerator.GroupKey> groupKeyIterator, int numKeys, int numFunctions,
       ColumnDataType[] resultStoredTypes) {
     GroupIdGenerator.GroupKey groupKey = groupKeyIterator.next();
     int groupId = groupKey._groupId;
@@ -293,60 +283,64 @@ public class MultistageGroupByExecutor {
 
   private void processAggregate(TransferableBlock block) {
     if (_maxFilterArgId < 0) {
-      // No filter for any aggregation function
-      int[] intKeys = generateGroupByKeys(block);
-      for (int i = 0; i < _aggFunctions.length; i++) {
-        AggregationFunction aggFunction = _aggFunctions[i];
-        Map<ExpressionContext, BlockValSet> blockValSetMap = AggregateOperator.getBlockValSetMap(aggFunction, block);
-        GroupByResultHolder groupByResultHolder = _aggregateResultHolders[i];
-        groupByResultHolder.ensureCapacity(_groupIdGenerator.getNumGroups());
-        aggFunction.aggregateGroupBySV(block.getNumRows(), intKeys, groupByResultHolder, blockValSetMap);
-      }
+      processAggregateWithoutFilter(block);
     } else {
-      // Some aggregation functions have filter, cache the matching rows
-      int[] intKeys = null;
-      RoaringBitmap[] matchedBitmaps = new RoaringBitmap[_maxFilterArgId + 1];
-      int[] numMatchedRowsArray = new int[_maxFilterArgId + 1];
-      int[][] filteredIntKeysArray = new int[_maxFilterArgId + 1][];
-      for (int i = 0; i < _aggFunctions.length; i++) {
-        AggregationFunction aggFunction = _aggFunctions[i];
-        int filterArgId = _filterArgIds[i];
-        if (filterArgId < 0) {
-          // No filter for this aggregation function
-          if (intKeys == null) {
-            intKeys = generateGroupByKeys(block);
-          }
-          Map<ExpressionContext, BlockValSet> blockValSetMap = AggregateOperator.getBlockValSetMap(aggFunction, block);
-          GroupByResultHolder groupByResultHolder = _aggregateResultHolders[i];
-          groupByResultHolder.ensureCapacity(_groupIdGenerator.getNumGroups());
-          aggFunction.aggregateGroupBySV(block.getNumRows(), intKeys, groupByResultHolder, blockValSetMap);
-        } else {
-          // Need to filter the block before aggregation
-          RoaringBitmap matchedBitmap = matchedBitmaps[filterArgId];
-          if (matchedBitmap == null) {
-            matchedBitmap = AggregateOperator.getMatchedBitmap(block, filterArgId);
-            matchedBitmaps[filterArgId] = matchedBitmap;
-            int numMatchedRows = matchedBitmap.getCardinality();
-            numMatchedRowsArray[filterArgId] = numMatchedRows;
-            filteredIntKeysArray[filterArgId] = generateGroupByKeys(block, numMatchedRows, matchedBitmap);
-          }
-          int numMatchedRows = numMatchedRowsArray[filterArgId];
-          int[] filteredIntKeys = filteredIntKeysArray[filterArgId];
-          Map<ExpressionContext, BlockValSet> blockValSetMap =
-              AggregateOperator.getFilteredBlockValSetMap(aggFunction, block, numMatchedRows, matchedBitmap);
-          GroupByResultHolder groupByResultHolder = _aggregateResultHolders[i];
-          groupByResultHolder.ensureCapacity(_groupIdGenerator.getNumGroups());
-          aggFunction.aggregateGroupBySV(numMatchedRows, filteredIntKeys, groupByResultHolder, blockValSetMap);
+      processAggregateWithFilter(block);
+    }
+  }
+
+  private void processAggregateWithoutFilter(TransferableBlock block) {
+    int[] intKeys = generateGroupByKeys(block);
+    int numGroups = _groupIdGenerator.getNumGroups();
+    for (int i = 0; i < _aggFunctions.length; i++) {
+      AggregationFunction aggFunction = _aggFunctions[i];
+      Map<ExpressionContext, BlockValSet> blockValSetMap = AggregateOperator.getBlockValSetMap(aggFunction, block);
+      GroupByResultHolder groupByResultHolder = _aggregateResultHolders[i];
+      groupByResultHolder.ensureCapacity(numGroups);
+      aggFunction.aggregateGroupBySV(block.getNumRows(), intKeys, groupByResultHolder, blockValSetMap);
+    }
+  }
+
+  private void processAggregateWithFilter(TransferableBlock block) {
+    // In the first loop, generate all the group keys, cache the matching rows
+    int[] intKeys = _filteredAggregationsSkipEmptyGroups ? null : generateGroupByKeys(block);
+    RoaringBitmap[] matchedBitmaps = new RoaringBitmap[_maxFilterArgId + 1];
+    int[] numMatchedRowsArray = new int[_maxFilterArgId + 1];
+    int[][] filteredIntKeysArray = new int[_maxFilterArgId + 1][];
+    for (int filterArgId : _filterArgIds) {
+      if (filterArgId < 0) {
+        // No filter for this aggregation function
+        if (intKeys == null) {
+          intKeys = generateGroupByKeys(block);
+        }
+      } else {
+        // Need to filter the block before aggregation
+        if (matchedBitmaps[filterArgId] == null) {
+          RoaringBitmap matchedBitmap = AggregateOperator.getMatchedBitmap(block, filterArgId);
+          matchedBitmaps[filterArgId] = matchedBitmap;
+          int numMatchedRows = matchedBitmap.getCardinality();
+          numMatchedRowsArray[filterArgId] = numMatchedRows;
+          filteredIntKeysArray[filterArgId] = generateGroupByKeys(block, numMatchedRows, matchedBitmap);
         }
       }
-      if (intKeys == null && !_filteredAggregationsSkipEmptyGroups) {
-        // _groupIdGenerator should still have all the groups even if there are only filtered aggregates for SQL
-        // compliant results. However, if the query option to skip empty groups is set, we avoid this step for
-        // improved performance.
-        generateGroupByKeys(block);
-        for (int i = 0; i < _aggFunctions.length; i++) {
-          _aggregateResultHolders[i].ensureCapacity(_groupIdGenerator.getNumGroups());
-        }
+    }
+
+    // In the second loop, aggregate the values
+    int numGroups = _groupIdGenerator.getNumGroups();
+    for (int i = 0; i < _aggFunctions.length; i++) {
+      AggregationFunction aggFunction = _aggFunctions[i];
+      GroupByResultHolder groupByResultHolder = _aggregateResultHolders[i];
+      groupByResultHolder.ensureCapacity(numGroups);
+      int filterArgId = _filterArgIds[i];
+      if (filterArgId < 0) {
+        Map<ExpressionContext, BlockValSet> blockValSetMap = AggregateOperator.getBlockValSetMap(aggFunction, block);
+        aggFunction.aggregateGroupBySV(block.getNumRows(), intKeys, groupByResultHolder, blockValSetMap);
+      } else {
+        Map<ExpressionContext, BlockValSet> blockValSetMap =
+            AggregateOperator.getFilteredBlockValSetMap(aggFunction, block, numMatchedRowsArray[filterArgId],
+                matchedBitmaps[filterArgId]);
+        aggFunction.aggregateGroupBySV(numMatchedRowsArray[filterArgId], filteredIntKeysArray[filterArgId],
+            groupByResultHolder, blockValSetMap);
       }
     }
   }

